ellipticcurve/utils/der.go

```diff
@@ -0,0 +1,210 @@
+	return EncodePrimitive(Sequence, strings.Join(encodedValues, ""))
+}
+
+func EncodePrimitive(tagType string, value interface{}) string {
+	if tagType == Integer {
+		value = EncodeInteger(value.(*big.Int))
+	}
+	if tagType == Object {
+		value = OidToHex(value.([]int64))
+	}
+	return fmt.Sprintf("%s%s%s", typeToHexTag[tagType], GenerateLengthBytes(value.(string)), value)
+}
+
+func Parse(hexadecimal string) []interface{} {
+	if hexadecimal == "" {
+		return nil
+	}
+	typeByte := hexadecimal[:2]
+	hexadecimal = hexadecimal[2:]
+	length, lengthBytes := ReadLengthBytes(hexadecimal)
+	content := hexadecimal[lengthBytes : lengthBytes+length]
+	hexadecimal = hexadecimal[lengthBytes+length:]
+	if len(content) < length {
+		panic("missing bytes in DER parse")
+	}
+
+	tagData := GetTagData(typeByte)
+	contentArray := make([]interface{}, len(content))
+	for i, value := range content {
+		contentArray[i] = value
+	}
+
+	if tagData["isConstructed"].(bool) {
+		nextContent := Parse(hexadecimal)
+		if len(nextContent) == 0 {
+			return []interface{}{Parse(content)}
+		}
+		return append([]interface{}{Parse(content)}, Parse(hexadecimal)[0])
+	}
+
+	switch tagData["type"] {
+	case Null:
+		contentArray = []interface{}{ParseNull(content)}
+	case Object:
+		contentArray = []interface{}{ParseOid(content)}
+	case UtcTime:
+		contentArray = []interface{}{ParseTime(content)}
+	case Integer:
+		contentArray = []interface{}{ParseInteger(content)}
+	case PrintableString:
+		contentArray = []interface{}{ParseString(content)}
+	default:
+		contentArray = []interface{}{ParseAny(content)}
+	}
+	return append(contentArray, Parse(hexadecimal)...)
+}
+
+func ParseAny(hexadecimal string) string {
+	return hexadecimal
+}
+
+func ParseOid(hexadecimal string) []int {
+	return OidFromHex(hexadecimal)
+}
+
+func ParseTime(hexadecimal string) time.Time {
+	parsedHex := ParseString(hexadecimal)
+	layout := "060102150405"
+	parsedTime, _ := time.Parse(layout, parsedHex)
+	return parsedTime
+}
+
+func ParseString(hexadecimal string) string {
+	return string(ByteStringFromHex(hexadecimal)[:])
+}
+
+func ParseNull(content string) string {
+	return ""
+}
+
+func ParseInteger(hexadecimal string) *big.Int {
+	integer := IntFromHex(hexadecimal)
+	bits := BitsFromHex(hexadecimal[0:1])
+	if bits[0:1] == "0" { // negative numbers are encoded using two's complement
+		return integer
+	}
+	bitCount := 4 * len(hexadecimal)
+	return integer.Sub(integer, big.NewInt(0).Exp(big.NewInt(2), big.NewInt(int64(bitCount)), nil))
+}
+
+func EncodeInteger(number *big.Int) string {
+	hexadecimal := HexFromInt(big.NewInt(0).Abs(number))
+	if number.Cmp(big.NewInt(0)) == -1 {
+		bitCount := big.NewInt(0).Mul(big.NewInt(4), big.NewInt(int64(len(hexadecimal))))
+		twosComplement := big.NewInt(0).Add(big.NewInt(0).Exp(big.NewInt(2), bitCount, nil), number)
+		return HexFromInt(twosComplement)
+	}
+	bits := BitsFromHex(string(hexadecimal[0]))
+	if string(bits[0]) == "1" {
```

Comment:
Pra ficar igual ao ecdsa-python, coloca esse comentário na frente:
"if first bit was left as 1, number would be parsed as a negative integer with two's complement"